pydis_site/apps/home/signals.py

```diff
@@ -0,0 +1,159 @@
+            return
+
+        self._apply_groups(user, account)
+
+    def social_account_removed(self, sender: Type[SocialLogin], **kwargs) -> None:
+        """Handle signals relating to removal of social accounts."""
+        account: SocialAccount = kwargs["socialaccount"]
+        provider: Provider = account.get_provider()
+
+        if not isinstance(provider, DiscordProvider):
+            return
+
+        try:
+            user: DiscordUser = DiscordUser.objects.get(id=int(account.uid))
+        except DiscordUser.DoesNotExist:
+            return
+
+        self._apply_groups(user, account, True)
+
+    def model_updated(self, sender: Type[DiscordUser], **kwargs) -> None:
+        """Handle signals related to the updating of Discord User model entries."""
+        instance: DiscordUser = kwargs["instance"]
+        raw: bool = kwargs["raw"]
+
+        # `update_fields` could be used for checking changes, but it's None here due to how the
+        # model is saved without using that argument - so we can't use it.
+
+        if raw:
+            # Fixtures are being loaded, so don't touch anything
+            return
+
+        try:
+            account: SocialAccount = SocialAccount.objects.get(
+                uid=str(instance.id), provider=DiscordProvider.id
+            )
+        except SocialAccount.DoesNotExist:
+            return  # User has never logged in with Discord on the site
+
+        self._apply_groups(instance, account)
+
+    def _apply_groups(
+            self, user: DiscordUser, account: SocialAccount, deletion: bool = False
+    ) -> None:
+        mappings = RoleMapping.objects.all()
+
+        try:
+            current_groups: List[Group] = list(account.user.groups.all())
+        except SocialAccount.user.RelatedObjectDoesNotExist:
+            return  # There's no user account yet, this will be handled by another receiver
+
+        if not user.in_guild:
+            deletion = True
+
+        if deletion:
+            # They've unlinked Discord or left the server, so we have to remove their groups
+
+            if not current_groups:
+                return  # They have no groups anyway, no point in processing
+
+            account.user.groups.remove(
+                *(mapping.group for mapping in mappings)
+            )
+        else:
+            new_groups = []
+
+            for role in user.roles.all():
+                try:
+                    new_groups.append(mappings.get(role=role).group)
+                except RoleMapping.DoesNotExist:
+                    continue  # No mapping exists
+
+            remove_groups = [
+                mapping.group for mapping in mappings if mapping.group not in new_groups
+            ]
+
+            add_groups = [group for group in new_groups if group not in current_groups]
+
+            if remove_groups:
+                account.user.groups.remove(*remove_groups)
+
+            if add_groups:
+                account.user.groups.add(*add_groups)
```

Comment:
Not fully convinced you need the `if` conditions here. I don't think it hurts to always run this and keep the code simpler